src/measure_theory/borel_hierarchy.lean

```diff
@@ -0,0 +1,440 @@
+values of the ordinal $\alpha$. Here the definition is extended in such a way that
+$\Sigma^0_0 = \emptyset$ (which results in more generality for some lemmas) and
+$\Pi^0_0$ coincides with the set of generators.
+
+The `inductive` definition of `sigma0_pi0_rec` was suggested by Junyan Xu.
+
+## References
+
+The definition of the hierarchy was extracted from [kechris1995].
+-/
+
+universe u
+
+namespace pointclasses
+
+section sigma0_pi0
+
+/-!
+### Σ and Π pointclasses
+
+This section includes basic definitions and API.
+-/
+open set
+
+variables {α : Type u} (s : set (set α)) (i k : ordinal.{u})
+
+/--
+Simultaneous recursive definition of Σ⁰ᵢ and Π⁰ᵢ pointclasses by recursion
+on ordinals (a variant of 11.B in Kechris, [Classical Descriptive Set Theory][kechris1995]).
+
+The definition is encoded as a single inductive predicate, where the `bool` argument
+determines if we are defining Σ⁰ᵢ (for `ff`) or Π⁰ᵢ (for `tt`). The parameter
+`s : set (set α)` is the generating family.
+
+The main difference with the classical definition is that this hierarchy starts
+at level 0: Π⁰₀ coincides with `s`
+augmented with `∅` and `univ` (intended to be a countable basis of a topology) and
+Σ⁰₀ is the empty family.
+-/
+inductive sigma0_pi0_rec {α : Type u} (s : set (set α)) :
+  ordinal.{u} → bool → set α → Prop
+| basic (x ∈ s) : sigma0_pi0_rec 0 tt x
+| empty         : sigma0_pi0_rec 0 tt ∅
+| univ          : sigma0_pi0_rec 0 tt univ
+| compl {i x}   : sigma0_pi0_rec i ff x → sigma0_pi0_rec i tt xᶜ
+| union {i} (f : ℕ → set α) (g : ℕ → ordinal.{u}) :
+    (∀ n, g n < i) → (∀ n, sigma0_pi0_rec (g n) tt (f n)) → sigma0_pi0_rec i ff (⋃ n, f n)
+
+/--
+The family of (boldface) Σ⁰ᵢ pointsets, which are countable unions of Π⁰ⱼ sets
+(given by the function `pointclasses.pi0` below) of smaller index. The parameter
+`s : set (set α)` is the family of sets from which the generation begins.
+-/
+def sigma0 : set (set α) := sigma0_pi0_rec s i ff
+
+/--
+The family of (boldface) Π⁰ᵢ pointsets, which are the complements of Σ⁰ᵢ sets
+(given by the function `pointclasses.sigma0` above). The parameter
+`s : set (set α)` is the family of sets from which the generation begins.
+
+When the ordinal argument is `0`, `pi0` returns the generating family `s`.
+-/
+def pi0 : set (set α) := sigma0_pi0_rec s i tt
+
+lemma sigma0_pi0_rec_def' {b : bool} :
+  sigma0_pi0_rec s i b = if b then pi0 s i else sigma0 s i :=
+by { unfold pi0 sigma0, cases b; refl }
+
+/-- Base case for `sigma0`. -/
+@[simp] lemma sigma0_zero : sigma0 s 0 = ∅ :=
+begin
+  unfold sigma0,
+  ext x, rw [mem_empty_iff_false, iff_false],
+  intro hx,
+  induction' hx with _ _ _ _ _ _ _ f g glt hf IH,
+  exact ordinal.not_lt_zero (g 0) (glt 0)
+ end
+
+/-- `sigma0` comprises unions of previous `pi0`. -/
+lemma sigma0_eq_Union_pi0:
+  sigma0 s i = set.range (λ (f : ℕ → ⋃ j (hij : j < i), pi0 s j), ⋃ n, (f n).1) :=
+begin
+  rcases classical.em (i=0) with rfl | hi; unfold sigma0, rw sigma0_pi0_rec_def',
```

Comment:
No need to separate the case `i = 0` from the other one here, the second proof works all the time. Then you can move the previous lemma after this one, and the proof just becomes `by simp [sigma0_eq_Union_pi0, ordinal.not_lt_zero]`.